gpt_index/optimzation/optimizer.py

```
@@ -0,0 +1,102 @@
+                    nltk.download("punkt")
+                tokenizer = nltk.data.load("tokenizers/punkt/english.pickle")
+                split_text = tokenizer.tokenize(textChunk.text)
+                if self._comparison_mode == "embedding":
+                    query_embedding = self._embed_model.get_agg_embedding_from_queries(
+                        query_bundle.embedding_strs
+                    )
+                    text_embeddings = self._embed_model._get_text_embeddings(split_text)
+                    num_top_k = None
+                    threshold = None
+                    if "percentile" in self._cutoffs:
+                        percentile = self._cutoffs["percentile"]
+                        num_top_k = int(len(split_text) * percentile)
+                    if "threshold" in self._cutoffs:
+                        threshold = self._cutoffs["threshold"]
+                    top_similarities, top_idxs = get_top_k_embeddings(
+                        query_embedding=query_embedding,
+                        embeddings=text_embeddings,
+                        similarity_fn=self._embed_model.similarity,
+                        similarity_top_k=num_top_k,
+                        embedding_ids=[i for i in range(len(text_embeddings))],
+                        similarity_cutoff=threshold,
+                    )
+                    if len(top_idxs) == 0:
+                        raise ValueError("Optimizer returned zero sentences.")
+                    top_sentences = [split_text[i] for i in top_idxs]
+
+                    logging.debug(f"> Top {len(top_idxs)} sentences with scores:\n")
+                    for i in range(len(top_idxs)):
+                        logging.debug(
+                            f"{i}. {top_sentences[i]} ({top_similarities[i]})"
+                        )
+                    return TextChunk(text=" ".join(top_sentences))
+
+            else:
+                print("Invalid split mode")
```

raise valueerror instead of printing

---

No longer applicable.